src/NUnitEngine/nunit.engine/Services/AgentLaunchers/AgentLauncherBase.cs

```diff
@@ -0,0 +1,124 @@
+        protected abstract TestAgentType AgentType { get; }
+        protected abstract FrameworkName TargetRuntime { get; }
+
+        protected abstract string AgentPath { get; }
+        protected virtual string X86AgentPath => throw new NotImplementedException(".NET Framework agents must override X86AgentPath");
+
+        public TestAgentInfo AgentInfo => new TestAgentInfo(AgentName, TestAgentType.LocalProcess, TargetRuntime);
+
+        public bool CanCreateAgent(TestPackage package)
+        {
+            // Get target runtime from package
+            string runtimeSetting = package.GetSetting(EnginePackageSettings.TargetFrameworkName, string.Empty);
+            var targetRuntime = new FrameworkName(runtimeSetting);
+
+            return targetRuntime.Identifier == TargetRuntime.Identifier && targetRuntime.Version.Major <= TargetRuntime.Version.Major;
+        }
+
+        public Process CreateAgent(Guid agentId, string agencyUrl, TestPackage package)
+        {
+            // Should not be called unless we have previously checked CanCreateAgent
+            Guard.ArgumentValid(CanCreateAgent(package), "Unable to create agent. Check result of CanCreateAgent before calling CreateAgent.", nameof(package));
+
+            var process = new Process()
+            {
+                EnableRaisingEvents = true
+            };
+
+            // Access package settings
+            bool runAsX86 = package.GetSetting(EnginePackageSettings.RunAsX86, false);
+            bool debugTests = package.GetSetting(EnginePackageSettings.DebugTests, false);
+            bool debugAgent = package.GetSetting(EnginePackageSettings.DebugAgent, false);
+            string traceLevel = package.GetSetting(EnginePackageSettings.InternalTraceLevel, "Off");
+            bool loadUserProfile = package.GetSetting(EnginePackageSettings.LoadUserProfile, false);
+            string workDirectory = package.GetSetting(EnginePackageSettings.WorkDirectory, string.Empty);
+
+            var sb = new StringBuilder($"--agentId={agentId} --agencyUrl={agencyUrl} --pid={Process.GetCurrentProcess().Id}");
+
+            // Set options that need to be in effect before the package
+            // is loaded by using the command line.
+            if (traceLevel != "Off")
+                sb.Append(" --trace=").EscapeProcessArgument(traceLevel);
+            if (debugAgent)
+                sb.Append(" --debug-agent");
+            if (debugTests)
+                sb.Append(" --debug-tests");
+            if (workDirectory != string.Empty)
+                sb.Append(" --work=").EscapeProcessArgument(workDirectory);
+
+            string arguments = sb.ToString();
+
+            var startInfo = process.StartInfo;
+            startInfo.UseShellExecute = false;
+            startInfo.CreateNoWindow = true;
+            startInfo.WorkingDirectory = Environment.CurrentDirectory;
+            startInfo.LoadUserProfile = loadUserProfile;
+
+            switch (TargetRuntime.Identifier)
```

Comment:
Why have this switch statement which is essentially logic specific to the subclasses of launcher instead of just isolating that logic to them.  Moving these out would also allow you to remove a bunch of stuff like AgentPath and AgentPathX86 out of this classe.

Reply:
Maybe move this into an abstract `PopulateStartInfo` method?

Reply:
@veleek 
For this commit, I'm not doing any of the above but I have simplified setting up StartInfo quite a bit. Tell me what you think.